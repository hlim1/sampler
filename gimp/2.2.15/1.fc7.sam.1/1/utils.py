from itertools import imap
from string import Template
from subprocess import PIPE, Popen

from SCons.Action import Action

import sys
sys.path[1:1] = ['/usr/lib/scons']


class AtTemplate(Template):

    delimiter = '@'

    pattern = r"""
    (?:
    (?P<escaped>@@) |
    (?P<named>(?!)) |
    @(?P<braced>%(id)s)@ |
    (?P<invalid>(?!))
    )
    """ % {'id': Template.idpattern}


def instantiate(source, sink, **kwargs):
    source = file(source)
    sink = file(sink, 'w')
    for line in source:
        sink.write(AtTemplate(line).substitute(kwargs))
    sink.close()


def read_pipe(action, env):

    def spawn(shell, escape, cmd, args, env):
        __pychecker__ = 'no-argsused'
        command = ' '.join(args)
        process = Popen(command, shell=True, executable=shell, env=env, stdout=PIPE)
        return process.stdout

    env = env.Copy(SPAWN=spawn)
    return env.Execute(action)


def __literal_exec(target, source, env):
    __pychecker__ = 'no-argsused'
    target = file(str(target[0]), 'w')
    print >>target, source[0].get_contents()
    target.close()

def __literal_show(target, source, env):
    return 'create "%s" containing %s' % (target[0], source[0])

__literal_action = Action(__literal_exec, __literal_show)


def literal(env, target, value):
    return env.Command(target, env.Value(value), __literal_action)
