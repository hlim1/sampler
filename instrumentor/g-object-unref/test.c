#include <glib-object.h>
#include <gtk/gtkmain.h>
#include <gtk/gtkwindow.h>


int main(int argc, char *argv[])
{
  GtkWindowGroup *object;

  gtk_init(&argc, &argv);

  object = gtk_window_group_new();
  g_object_ref(object);

  g_object_unref(object);
  g_object_unref(object);
  g_object_unref(object);

  return 0;
}
