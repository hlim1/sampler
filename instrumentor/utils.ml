open Cil
open Pretty

let instr_what = function
  | Set(_, _, location) -> "Set"
  | Call(_, _, _, location) -> "Call"
  | Asm(_, _, _, _, _, location) -> "Asm"

let stmt_what = function
  | Instr(instrs) -> Printf.sprintf "Instr × %i" (List.length instrs)
  | Return(_) -> "Return"
  | Goto(_) -> "Goto"
  | Break(_) -> "Break"
  | Continue(_) -> "Continue"
  | If(_) -> "If"
  | Switch(_) -> "Switch"
  | Loop(_) -> "Loop"
  | Block({bstmts = bstmts}) -> Printf.sprintf "Block × %i" (List.length bstmts)
  | TryFinally(_) -> "TryFinally"
  | TryExcept(_) -> "TryExcept"

let stmt_describe stmt =
  let where = get_stmtLoc stmt in
  Printf.sprintf "%s:%i: %s" where.file where.line (stmt_what stmt)
    
let d_stmt _ stmt =
  dprintf "%a: CFG #%i: %s" d_loc (get_stmtLoc stmt.skind) stmt.sid (stmt_what stmt.skind)
    
let d_stmts _ stmts =
  seq line (d_stmt ()) stmts
    
let print_stmts stmts =
  fprint stdout 80 (d_stmts () stmts)
    
let warn stmt message =
  ignore(fprintf stderr "%a: %s\n" d_stmt stmt message)


let d_labels () {labels = labels} =

  let rec labelNames = function
  | [] -> []
  | Label (name, _, _) :: rest ->
      name :: labelNames rest
  | _ :: rest ->
      labelNames rest
  in

  chr '['
    ++ (seq
	  ~sep:(text "; ")
	  ~doit:text
	  ~elements:(labelNames labels))
    ++ chr ']'


let d_preds () {preds = preds} =
  seq
    ~sep:(text "; ")
    ~doit:(fun stmt -> num stmt.sid)
    ~elements:preds


let d_succs () {succs = succs} =
  seq
    ~sep:(text "; ")
    ~doit:(fun stmt -> num stmt.sid)
    ~elements:succs
