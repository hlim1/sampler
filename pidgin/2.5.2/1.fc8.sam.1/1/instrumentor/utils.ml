open Cil
open Pretty

let instr_what = function
  | Set _ -> "Set"
  | Call _ -> "Call"
  | Asm _ -> "Asm"

let stmt_what = function
  | Instr instrs -> Printf.sprintf "Instr × %i" (List.length instrs)
  | Return _ -> "Return"
  | Goto _ -> "Goto"
  | Break _ -> "Break"
  | Continue _ -> "Continue"
  | If _ -> "If"
  | Switch _ -> "Switch"
  | Loop _ -> "Loop"
  | Block {bstmts = bstmts} -> Printf.sprintf "Block × %i" (List.length bstmts)
  | TryFinally _ -> "TryFinally"
  | TryExcept _ -> "TryExcept"

let stmt_describe stmt =
  let where = get_stmtLoc stmt in
  Printf.sprintf "%s:%i: %s" where.file where.line (stmt_what stmt)
    
let d_stmt _ stmt =
  dprintf "%a: CFG #%i: %s" d_loc (get_stmtLoc stmt.skind) stmt.sid (stmt_what stmt.skind)
    
let d_stmts _ stmts =
  seq line (d_stmt ()) stmts
    
let print_stmts stmts =
  fprint stdout 80 (d_stmts () stmts)
    
let warn stmt message =
  ignore(fprintf stderr "%a: %s\n" d_stmt stmt message)


let d_label () = function
  | Label (name, _, _) ->
      text name
  | Case (expr, _) ->
      text "case " ++ d_exp () expr
  | Default _ ->
      text "default"

let d_labels () labels =
  seq
    ~sep:(text "; ")
    ~doit:(d_label ())
    ~elements:labels


let d_preds () {preds = preds} =
  seq
    ~sep:(text "; ")
    ~doit:(fun stmt -> num stmt.sid)
    ~elements:preds


let d_succs () {succs = succs} =
  seq
    ~sep:(text "; ")
    ~doit:(fun stmt -> num stmt.sid)
    ~elements:succs
