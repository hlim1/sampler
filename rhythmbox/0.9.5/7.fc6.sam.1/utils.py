from itertools import imap
from string import Template
from subprocess import PIPE, Popen

import sys
sys.path[1:1] = ['/usr/lib/scons']


class AtTemplate(Template):

    delimiter = '@'

    pattern = r"""
    (?:
    (?P<escaped>@@) |
    (?P<named>(?!)) |
    @(?P<braced>%(id)s)@ |
    (?P<invalid>(?!))
    )
    """ % {'id': Template.idpattern}


def instantiate(source, sink, **kwargs):
    source = file(source)
    sink = file(sink, 'w')
    for line in source:
        sink.write(AtTemplate(line).substitute(kwargs))
    sink.close()


def read_pipe(action, env):

    def spawn(shell, escape, cmd, args, env):
        __pychecker__ = 'no-argsused'
        command = ' '.join(args)
        process = Popen(command, shell=True, executable=shell, env=env, stdout=PIPE)
        return process.stdout

    env = env.Copy(SPAWN=spawn)
    return env.Execute(action)


def __literal_action(target, source, env):
    __pychecker__ = 'no-argsused'
    target = file(str(target[0]), 'w')
    print >>target, source[0].get_contents()
    target.close()


def literal(env, target, value):
    return env.Command(target, env.Value(value), __literal_action)
