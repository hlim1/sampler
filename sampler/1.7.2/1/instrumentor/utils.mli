open Cil
open Pretty


val instr_what : instr -> string
val stmt_what : stmtkind -> string
val stmt_describe : stmtkind -> string

val d_stmt : unit -> stmt -> doc
val d_stmts : unit -> stmt list -> doc
val d_label : unit -> label -> doc
val d_labels : unit -> label list -> doc
val d_preds : unit -> stmt -> doc
val d_succs : unit -> stmt -> doc

val print_stmts : stmt list -> unit
val warn : stmt -> string -> unit
