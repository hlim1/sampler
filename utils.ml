open Cil
open Pretty

let instr_where = function
  | Set(_, _, location) -> location
  | Call(_, _, _, location) -> location
  | Asm(_, _, _, _, _, location) -> location

let instr_what = function
  | Set(_, _, location) -> "Set"
  | Call(_, _, _, location) -> "Call"
  | Asm(_, _, _, _, _, location) -> "Asm"

let rec stmt_where = function
  | Instr([]) -> locUnknown
  | Instr(head::_) -> instr_where head
  | Return(_, location) -> location
  | Goto(_, location) -> location
  | Break(location) -> location
  | Continue(location) -> location
  | If(_, _, _, location) -> location
  | Switch(_, _, _, location) -> location
  | Loop(_, location) -> location
  | Block({bstmts = []}) -> locUnknown
  | Block({bstmts = {skind = skind} :: _}) -> stmt_where skind

let stmt_what = function
  | Instr(instrs) -> Printf.sprintf "Instr × %i" (List.length instrs)
  | Return(_) -> "Return"
  | Goto(_) -> "Goto"
  | Break(_) -> "Break"
  | Continue(_) -> "Continue"
  | If(_) -> "If"
  | Switch(_) -> "Switch"
  | Loop(_) -> "Loop"
  | Block({bstmts = bstmts}) -> Printf.sprintf "Block × %i" (List.length bstmts)

let stmt_describe stmt =
  let where = stmt_where stmt in
  Printf.sprintf "%s:%i: %s" where.file where.line (stmt_what stmt)
    
let d_stmt _ stmt =
  dprintf "%a: CFG #%i: %s" d_loc (stmt_where stmt.skind) stmt.sid (stmt_what stmt.skind)
    
let d_stmts _ stmts =
  seq line (d_stmt ()) stmts
    
let print_stmts stmts =
  fprint stdout 80 (d_stmts () stmts)
    
let warn stmt message =
  ignore(fprintf stderr "%a: %s\n" d_stmt stmt message)
