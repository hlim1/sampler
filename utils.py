from itertools import imap
from string import Template
from subprocess import PIPE, Popen

import sys
sys.path[1:1] = ['/usr/lib/scons']


def read_pipe(action, env):

    def spawn(shell, escape, cmd, args, env):
        __pychecker__ = 'no-argsused'
        command = ' '.join(args)
        process = Popen(command, shell=True, executable=shell, env=env, stdout=PIPE)
        return process.stdout

    env = env.Copy(SPAWN=spawn)
    return env.Execute(action)


def instantiate(source, sink, **kwargs):
    source = file(source)
    sink = file(sink, 'w')

    for line in source:
        sink.write(Template(line).substitute(kwargs))


def __literal_action(target, source, env):
    __pychecker__ = 'no-argsused'
    target = file(str(target[0]), 'w')
    print >>target, source[0].get_contents()
    target.close()


def literal(env, target, value):
    return env.Command(target, env.Value(value), __literal_action)
