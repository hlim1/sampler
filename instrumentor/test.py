import sys
sys.path[1:1] = ['/usr/lib/scons']

from SCons.Action import Action
from SCons.Builder import Builder
from SCons.Defaults import CScan, Touch
from SCons.Scanner import Scanner
from SCons.Script import Flatten

from itertools import chain, ifilter, imap

import filecmp

from utils import read_pipe


########################################################################
#
#  common utilities
#


__driver_deps = ['#driver/sampler-cc-here', '#driver/as', '#driver/cc1', '#driver/sampler-specs', '#instrumentor/main']


def __sampler_cc_scan(node, env, path):
    __pychecker__ = 'no-argsused'
    # todo: dependencies on headers under '#lib'
    return map(env.File, __driver_deps)


def __sampler_ld_scan(node, env, path):
    __pychecker__ = 'no-argsused'
    tdir = {True: 'yes', False: 'no'}['-pthread' in env['LINKFLAGS']]
    libs = ['early', 'late', 'random-online']
    libs = ( 'threads/%s/lib%s' % (tdir, lib) for lib in libs )
    libs = chain(['schemes/libschemes'], libs)
    libs = ( '#lib/%s.a' % lib for lib in libs )

    deps = chain(__driver_deps, libs)
    return map(env.File, deps)


__sampler_cc_scanner = Scanner(function=__sampler_cc_scan)
__sampler_ld_scanner = Scanner(function=__sampler_ld_scan)


########################################################################
#
#  static objects and programs
#


__static_object_builder = Builder(
    action='$CCCOM',
    prefix='$OBJPREFIX',
    suffix='$OBJSUFFIX',
    src_suffix='$CFILESUFFIX',
    target_scanner=__sampler_cc_scanner,
    )


__program_builder = Builder(
    action='$LINKCOM',
    prefix='$PROGPREFIX',
    suffix='$PROGSUFFIX',
    src_suffix='$OBJSUFFIX',
    src_builder='CBIStaticObject',
    target_scanner=__sampler_ld_scanner,
    )


########################################################################
#
#  shared objects and libraries
#


__shared_object_builder = Builder(
    action='$SHCCCOM',
    prefix='$SHOBJPREFIX',
    suffix='$SHOBJSUFFIX',
    src_suffix='$CFILESUFFIX',
    target_scanner=__sampler_cc_scanner,
    )


__shared_library_builder = Builder(
    action='$SHLINKCOM',
    prefix='$SHLIBPREFIX',
    suffix='$SHLIBSUFFIX',
    src_suffix='$SHOBJSUFFIX',
    src_builder='CBISharedObject',
    target_scanner=__sampler_ld_scanner,
    )


########################################################################
#
#  static information extraction
#


def __extract_scan(node, env, path):
    __pychecker__ = 'no-argsused'
    return [env.File('#$EXTRACT_SECTION')]


__extract_scanner = Scanner(function=__extract_scan)


__sites_action = Action([['$EXTRACT_SECTION', '.debug_site_info', '$SOURCES', '>$TARGET']])


__sites_builder = Builder(
    action=__sites_action,
    suffix='.sites',
    src_suffix='$PROGSUFFIX',
    src_builder='CBIProgram',
    target_scanner=__extract_scanner,
    )


########################################################################
#
#  feedback report creation
#


def __var_test_stdout(target, source, env, for_signature):
    __pychecker__ = 'no-argsused'
    if env['TEST_STDOUT']:
        return '>$TEST_STDOUT'


def __var_test_stderr(target, source, env, for_signature):
    __pychecker__ = 'no-argsused'
    if env['TEST_STDERR']:
        return '2>$TEST_STDERR'


__reports_action = Action([['{', '$TEST_PREFIX', 'env', 'SAMPLER_SPARSITY=1', 'SAMPLER_FILE=$TARGET', '$SOURCE', '$TEST_ARGS', ';', '}', '$_TEST_STDOUT', '$_TEST_STDERR']])


__reports_builder = Builder(
    action=__reports_action,
    suffix='.reports',
    src_suffix=[''],
    src_builder='CBIProgram',
    )


########################################################################
#
#  samples resolving
#


__resolved_samples_action = Action([[
    '$RESOLVE_SAMPLES', '$objects', '<$SOURCE', '|',
    'cut', '-f1,3-', '|',
    'sed', 's:$SOURCE.dir/::g', '|',
    'sort', '-t', '\t', '-o', '$TARGET',
    ]])


def __resolved_samples_scan(node, env, path):
    __pychecker__ = 'no-argsused'
    return Flatten([env.subst('#$RESOLVE_SAMPLES'), env['objects']])


__resolved_samples_scanner = Scanner(function=__resolved_samples_scan)


__resolved_samples_builder = Builder(
    action=__resolved_samples_action,
    suffix='.resolved',
    src_suffix='.reports',
    src_builder='CBIReports',
    target_scanner=__resolved_samples_scanner,
    )


########################################################################
#
#  timestamps resolving
#


__resolved_timestamps_action = Action([[
    '$RESOLVE_TIMESTAMPS', '${SOURCE.children()}', '<$SOURCE', '|',
    'cut', '-f1,4-', '|',
    'sed', 's:$SOURCE.dir/::g', '|',
    'sort', '-t', '\t', '-k3', '-o', '$TARGET',
    ]])


def __resolved_timestamps_scan(node, env, path):
    __pychecker__ = 'no-argsused'
    return [env.subst('#$RESOLVE_TIMESTAMPS')]


__resolved_timestamps_scanner = Scanner(function=__resolved_timestamps_scan)


__resolved_timestamps_builder = Builder(
    action=__resolved_timestamps_action,
    suffix='.timestamps',
    src_suffix='.reports',
    src_builder='CBIReports',
    target_scanner=__resolved_timestamps_scanner,
    )


########################################################################
#
#  control flow graph resolving
#


__resolved_cfg_action = Action([[
    '$RESOLVE_CFG', '$SOURCE', '|',
    'sed', 's:$SOURCE.dir/::g', '>$TARGET',
    ]])


def __resolved_cfg_scan(node, env, path):
    __pychecker__ = 'no-argsused'
    return [env.subst('#$RESOLVE_CFG')]


__resolved_cfg_scanner = Scanner(function=__resolved_cfg_scan)


__resolved_cfg_builder = Builder(
    action=__resolved_cfg_action,
    suffix='.cfg',
    src_suffix=['$OBJSUFFIX', '$SHOBJSUFFIX', '$LIBSUFFIX', '$SHLIBSUFFIX', '$PROGSUFFIX'],
    src_builder='CBIStaticObject',
    target_scanner=__resolved_cfg_scanner,
    )


########################################################################
#
#  comparison with reference output
#


def __compare_action_exec(target, source, env):
    __pychecker__ = 'no-argsused'
    [source] = source
    expected = source.target_from_source('', '.expected')
    return not filecmp.cmp(str(source), str(expected), False)


def __compare_action_show(target, source, env):
    __pychecker__ = 'no-argsused'
    [source] = source
    expected = source.target_from_source('', '.expected')
    return 'compare "%s" and "%s"' % (source, expected)


__compare_action = Action(__compare_action_exec, __compare_action_show)


__expect_action = [__compare_action, Touch('$TARGET')]


def __expect_scan(node, env, path):
    __pychecker__ = 'no-argsused'
    return [node.target_from_source('', '.expected')]


__expect_scanner = Scanner(function=__expect_scan)


__expect_builder = Builder(
    action=__expect_action,
    suffix='.passed',
    target_scanner=__expect_scanner,
    single_source=True,
    )


########################################################################


def generate(env):
    env['CC'] = env.File('#driver/sampler-cc-here')

    env.AppendUnique(
        CCFLAGS=['--relative-paths'],
        EXTRACT_SECTION=env.File('#tools/extract-section'),
        RESOLVE_CFG=env.File('#tools/resolve-cfg'),
        RESOLVE_SAMPLES=env.File('#tools/resolve-samples'),
        RESOLVE_TIMESTAMPS=env.File('#tools/resolve-timestamps'),

        BUILDERS={
        'CBIProgram': __program_builder,
        'CBIReports': __reports_builder,
        'CBIResolvedCFG': __resolved_cfg_builder,
        'CBIResolvedSamples': __resolved_samples_builder,
        'CBIResolvedTimestamps': __resolved_timestamps_builder,
        'CBISharedLibrary': __shared_library_builder,
        'CBISharedObject': __shared_object_builder,
        'CBISites': __sites_builder,
        'CBIStaticObject': __static_object_builder,
        'Expect': __expect_builder,
        },

        TEST_PREFIX=[],
        TEST_STDOUT=[],
        TEST_STDERR=[],
        _TEST_STDOUT='${_concat(">", TEST_STDOUT, "", __env__)}',
        _TEST_STDERR='${_concat("2>", TEST_STDERR, "", __env__)}',
        )


def exists(env):
    return True
