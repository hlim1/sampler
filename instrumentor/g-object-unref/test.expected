test	g-object-unref	test.c	28	test_finalize	2	(void *)self	1	0	0	0
test	g-object-unref	test.c	75	main	8	(void *)object	0	0	1	0
test	g-object-unref	test.c	76	main	12	(void *)object	0	1	0	0
